nexus/src/app/rack.rs

```diff
@@ -59,4 +62,53 @@ impl super::Nexus {
             Err(Error::not_found_by_id(ResourceType::Rack, rack_id))
         }
     }
+
+    /// Marks the rack as initialized with a set of services.
+    ///
+    /// This function is a no-op if the rack has already been initialized.
+    pub async fn rack_initialize(
+        &self,
+        opctx: &OpContext,
+        rack_id: Uuid,
+        services: Vec<ServicePutRequest>,
+    ) -> Result<(), Error> {
+        opctx.authorize(authz::Action::Modify, &authz::FLEET).await?;
+
+        let (.., db_rack) = LookupPath::new(opctx, &self.db_datastore)
+            .rack_id(rack_id)
+            .fetch()
+            .await?;
+
+        // TODO-concurrency: Technically, this initialization is not atomic.
+        // We check if the rack is initialized, upsert each service, then
+        // set the rack to be initialized.
+        //
+        // However, this should practically be safe for a few reasons:
+        //
+        // 1. There should only be a single RSS process executing at once.
+        // 2. Once the RSS propagates information about services to Nexus,
+        // they should be recorded to a local plan that does not change.
```

Comment:
RSS records its plan locally before propagating it to nexus, right? If it happened after, I'd have a question about whether a sequence like this were possible:

1. RSS determines a plan
2. RSS sends the plan to nexus
3. Something goes wrong partway through
4. RSS starts over and determines a different plan

where step 3 left bogus services from the initial plan in the db. Is there a downside to putting this in a transaction now?

Reply:
As discussed on the RFD: https://github.com/oxidecomputer/rfd/pull/434#discussion_r888200485 if the call here failed, I think we'd re-try using the same plan. 

That being said, I went ahead and moved all these calls into a transaction - this is a low-frequency call; it's probably worth the cost to not worry about safety issues.